trunk/doc/man/sxstr_mngr.3
.\" @(#)sxstr_mngr.3	- SYNTAX [unix] - 29 Decembre 1987
.TH SXSTR_MNGR 3 "SYNTAX\[rg]"
.SH NAME
sxstrmngr_t,
sxstrmngr,
sxstr_mngr,
sxstr_retrievelen,
sxstr_retrieve,
sxstr_savelen,
sxstr_save,
sxstr_get,
sxstr_length,
sxstr_size,
sxstr_dump,
sxstr_save_keywords,
sxstr2retrieve,
sxstr2save,
sxstrretrieve,
sxstrsave,
sxstrget,
sxstrlen,
SXSTRtop
\- string manager for SYNTAX.
.SH SYNOPSIS
.nf
.ta \w'void  'u +\w'void  'u
.B
#include "sxunix.h"
.PP
.B
typedef struct sxstrmngr sxstrmngr_t\|;
.PP
.B
void sxstr_mngr (what, strtable)
.B
	int	 what\|;
.B
	sxstrmngr_t	 *strtable\|;
.PP
General primitives operating on any string table passed as 1st argument
.PP
.B
unsigned int sxstr_save (strtable, string)
.B
	sxstrmngr_t	 *strtable\|;
.B
	char	*string\|;
.PP
.B
unsigned int sxstr_savelen (strtable, string, strlength)
.B
	sxstrmngr_t	 *strtable\|;
.B
	char	*string\|;
.B
	unsigned long  strlength\|;
.PP
.B
unsigned int sxstr_retrieve (strtable, string)
.B
	sxstrmngr_t	 *table\|;
.B
	char	*string\|;
.PP
.B
unsigned int sxstr_retrievelen (strtable, string, strlength)
.B
	sxstrmngr_t	 *table\|;
.B
	char	*string\|;
.B
	unsigned long  strlength\|;
.PP
.B
char *sxstr_get (strtable, string_table_entry)
.B
	sxstrmngr_t	 *strtable\|;
.B
	unsigned int  string_table_entry\|;
.PP
.B
unsigned long sxstr_length (strtable, string_table_entry)
.B
	sxstrmngr_t	 *strtable\|;
.B
	unsigned int  string_table_entry\|;
.PP
.B
unsigned int sxstr_size (strtable)
.B
	sxstrmngr_t	 *strtable\|;
.PP
.B
void sxstr_dump (f, strtable)
.B
	FILE	 *f\|;
.B
	sxstrmngr_t	 *strtable\|;
.PP
.B
void sxstr_save_keywords (strtable, lang)
.B
	sxstrmngr_t	*strtable\|;
.B
	char	*lang\|;
.PP
Traditional primitives that operate only on the global variable sxstrmngr
.PP
.B
sxstrmngr_t sxstrmngr\|;
.PP
.B
char *sxstrget (string_table_entry)
.B
	unsigned int  string_table_entry\|;
.PP
.B
unsigned long sxstrlen (string_table_entry)
.B
	unsigned int  string_table_entry\|;
.PP
.B
unsigned int sxstrsave (string)
.B
	char	*string\|;
.PP
.B
unsigned int sxstr2save (string, strlen)
.B
	char	*string\|;
.B
	unsigned long  strlen\|;
.PP
.B
unsigned int sxstrretrieve (string)
.B
	char	*string\|;
.PP
.B
unsigned int sxstr2retrieve (string, strlen)
.B
	char	*string\|;
.B
	unsigned long  strlen\|;
.PP
.B
unsigned int SXSTRtop ()
.fi
.SH DESCRIPTION
The
.B string manager
module allows to store strings once and subsequently refer to them
with a unique positive integral number (``string table entry'').  This
module provides two sets of primitives (procedures and
macro-definitions): the first set operates on any string table passed
as first argument, whereas the second set operates only on the default
string table, stored in the global variable
.I sxstrmngr
(of type
.IR \s-2sxstrmngr_t\s0 ,
the details of which are of no interest to the naive user).
The
.B SYNTAX
scanner (see
.BR sxscanner (3))
is a good example of the use of the second set of primitives.
.LP
.I sxstr_mngr
is the first and last entry point of the module to be called for using
a string table.
It may be called with four possible arguments\|:
.IP -
.IR \s-2SXBEGIN\s0
states that the default string table
.I sxstrmngr
should become initialized. If this table was already initialized
and used, the call
.I sxstr_mngr\ (\s-2SXBEGIN\s0)
reinitializes the table, the internal structures of which are
cleared, without the need to call
.I sxstr_mngr\ (\s-2SXEND\s0)
beforehand; in such case, it is possible and recommended to call
.I sxstr_mngr\ (\s-2SXCLEAR\s0)
rather than
.I sxstr_mngr\ (\s-2SXBEGIN\s0)
so as to better document the expected effect.
.IP -
.IR \s-2SXEND\s0
states that the strings saved in the default table
.I sxstrmngr
will be no more accessed, so that internal structures may be freed.
.IP -
.I \s-2SXOPEN\s0
.IR strtable
states that the string table
.I *strtable
should become initialized. If this table was already initialized and
used, the call
.I sxstr_mngr\ (\s-2SXOPEN\s0, strtable)
reinitializes
.IR *strtable,
without the need to call
.I sxstr_mngr\ (\s-2SXCLOSE\s0, strtable)
beforehand.
.IP -
.I \s-2SXCLOSE\s0
.IR strtable
statea that the strings saved in the table
.I *strtable
will be no more accessed, so that internal structures may be freed.
.PP
.I sxstr_get
returns (a pointer to) the null-terminated string associated with
.IR string_table_entry
in the table
.IR strtable .
.I sxstr_length
returns the number of characters in that string, not including the
terminating null character.
.I sxstr_get
and
.I sxstr_length
are (side-effect free) macros.
.PP
.I sxstr_save
saves the string
.I string
in the string table
.I strtable
and returns the associated unique number.
Further calls with a string comparing equal with
.I string
will return the same number.
.PP
.I sxstr_savelen
is similar\|; the difference is that the number of characters in
.I string
is given (as
.IR strlength),
and does not have to be computed.
This allows to save strings that contain null characters and may be
not null-terminated.
.PP
.I sxstr_retrieve
is only different from
.I sxstr_save
in that it does not store the string
.I string
in the string table if it is not there, and returns instead the
constant
.I \s-2SXERROR_STE\s0
(see \s-2NOTES\s0 below).
.PP
.I sxstr_retrievelen
is similar\|; the difference is that the number of characters in
.I string
is given (as
.IR strlength),
and does not have to be computed.
.PP
.I sxstr_size
is a side-effect-free macro that returns the current number of elements
(including the two special entries
.I \s-2SXERROR_STE\s0
and
.IR \s-2SXERROR_STE\s0 )
stored in the string table
.IR strtable .
.PP
.I sxstr_dump
prints the contents of the string table
.I strtable
to file
.IR f ,
which has to be opened before calling
.IR sxstr_dump .
.PP
.I sxstr_save_keywords
adds all the reserved keywords of the language
.I lang
to
.IR strtable.
Currently, only the C programming language is supported (i.e.,
.I lang
= "C").
.PP
The primitives of the second set of primitives differ from those of
the first set only by the fact that they operate on the default string
table, i.e., the global variable
.IR sxstrmngr :
.IP -
.I sxsxtrget (...)
is similar to
.I sxstr_get (&sxstrmngr, ...)
.IP -
.I sxsxtrlen (...)
is similar to
.I sxstr_length (&sxstrmngr, ...)
.IP -
.I sxsxtrsave (...)
is similar to
.I sxstr_save (&sxstrmngr, ...)
.IP -
.I sxsxtr2save (...)
is similar to
.I sxstr_savelen (&sxstrmngr, ...)
.IP -
.I sxsxtrretrieve (...)
is similar to
.I sxstr_retrieve (&sxstrmngr, ...)
.IP -
.I sxsxtr2retrieve (...)
is similar to
.I sxstr_retrievelen (&sxstrmngr, ...) 
.IP -
.I SXSTRtop()
is similar to
.I sxstr_size (&sxstrmngr)
.SH "SEE ALSO"
sxunix(3)
and the \fISYNTAX Reference Manual\fP.
.SH NOTES
Two special constants are defined in the header file \fI"sxunix.h"\fP.
.I \s-2SXERROR_STE\s0
is the string table entry associated with any generic token inserted
by the parser during error recovery.
.I \s-2SXEMPTY_STE\s0
is predefined as the string table entry associated with the empty
string.

.SH BUG
Because
.I sxstrget
and
.I sxstrlen
are implemented as macros, they should not be used directly with the
result of a call to
.I sxstrsave
or
.IR sxstr2save ,
as that call may change the value of variables used in the macros,
which might on some systems produce invalid results.
Thus the user is urged not to write things as\|:
.nr 53 \w'\fBste \fP'
.if \n(53<\w'\fBptr \fP' .nr 53 \w'\fBptr \fP'
.ta \w'    'u +\n(53u
.br
\fB	ptr	= sxstrget (sxstrsave (string))\|;\fP
.br
and to use an intermediate variable instead, as in\|:
.nf
\fB	ste	= sxstrsave (string)\|;\fP
\fB	ptr	= sxstrget (ste)\|;\fP
.fi



trunk/src/sxstr_mngr.c
/******************************************************************************
 *                                S Y N T A X
 *-----------------------------------------------------------------------------
 *   Copyright (C) 1972-2023 INRIA (Institut National de Recherche en
 *   Informatique et Automatique)
 *-----------------------------------------------------------------------------
 *   URL: http://sourcesup.renater.fr/projects/syntax
 *-----------------------------------------------------------------------------
 *   The source code of SYNTAX is distributed with two different licenses,
 *   depending on the files:
 *   - The recursive content of src/ and incl/ and the non-recursive content
 *     of SYNTAX's root directory are distributed under the CeCILL-C license
 *   - The recursive content of all other repertories is distributed under
 *     the CeCILL license
 *   All code produced by SYNTAX must be considered as being under the
 *   CeCILL-C license. Information about the CeCILL and CeCILL-C licenses
 *   can be found at, e.g., http://www.cecill.info
 *****************************************************************************/

/* Sauvegarde des chaines de caracteres */

#include "sxversion.h"
#include "sxunix.h"
#include <stdarg.h>

char WHAT_SXSTR_MNGR[] = "@(#)SYNTAX - $Id: sxstr_mngr.c 3633 2023-12-20 18:41:19Z garavel $" WHAT_DEBUG;

/*---------------------------------------------------------------------------*/

#include "sxscrmbl.h"

#if (SXHASH_LENGTH % 2) == 0 || (SXHASH_LENGTH % 3) == 0 || (SXHASH_LENGTH % 5) == 0 || (SXHASH_LENGTH % 7) == 0 || (SXHASH_LENGTH % 11) == 0 || (SXHASH_LENGTH % 13) == 0
#error SXHASH_LENGTH should not be divisible by a small prime.  You might use one of {257,521,1031,2053,4099,8209,16411,32771,65539,131101,262217,524309,1048583,2097169} instead
#endif

/*---------------------------------------------------------------------------*/

/* Increment for the string reallocation */

#ifndef BUFSIZ
#define	BUFSIZ	512
#endif	/* BUFSIZ */

/*---------------------------------------------------------------------------*/

/* Compare for equality strings s1 and s2 on their first n characters. */
/* The strings may contain '\0's and may not be terminated by a '\0'. */

static SXINLINE bool sxstrequal (char *s1, char *s2, SXUINT n)
{
    do
	if (n-- == 0)
	    return true;
    while (*s1++ == *s2++);

    return false;
}

/*---------------------------------------------------------------------------*/

/* Put in s1 first (n-1) chars of s2 and terminate by a '\0' */

static SXINLINE void sxstrput (char *s1, char *s2, SXUINT n)
{
    while (--n != 0)
	*s1++ = *s2++;

    *s1 = SXNUL;
}

/*---------------------------------------------------------------------------*/

/* Enter into strtable a string of length strlength for entry top++ */

static SXINT sxstr_add (sxstrmngr_t *strtable, char *string, SXUINT strlength)
{
    SXUINT	length = strlength;
    SXINT	ste;

    if (strtable->strnext + ++length >= strtable->strlength) {
	sxstrstring_t	*old;

	old = strtable->string;

#ifdef SXRESIZE
	SXRESIZE (old, 1, sizeof (sxstrstring_t*) + (strtable->strnext - 1) * sizeof (char));
	strtable->strlength += (length & -sizeof (char*)) + BUFSIZ;
#else
	/* This might be a buddy system allocator; find a power of 2 */

	if (strtable->strlength < length) {
	    SXUINT	min = length + sizeof (sxstrstring_t*) / sizeof (char);

	    strtable->strlength += sizeof (sxstrstring_t*) / sizeof (char);

	    do {
	    } while ((strtable->strlength *= 2) < min);

	    strtable->strlength -= sizeof (sxstrstring_t*) / sizeof (char);
	}
#endif	/* SXRESIZE */

	strtable->strnext = 0;
	strtable->string = (sxstrstring_t*) sxalloc (1, sizeof (sxstrstring_t*) + strtable->strlength * sizeof (char));
	strtable->string->next = old;
    }

    if ((ste = strtable->top++) >= (SXINT)strtable->tablength)
	strtable->table = (sxstrtable_t*) sxrealloc (strtable->table, strtable->tablength += SXHASH_LENGTH, sizeof (sxstrtable_t));

    {
	sxstrtable_t	*ptable = &(strtable->table [ste]);

	sxstrput (ptable->pointer = strtable->string->chars + strtable->strnext, string, length);
	strtable->strnext += length;
	ptable->length = strlength;
	ptable->collision_link = SXERROR_STE;
    }

    return ste;
}

/*---------------------------------------------------------------------------*/

/* Save a string of known length in strtable and return a unique identifier for it */
/* The string may be not terminated by a SXNUL and may contain SXNULs */

SXINT sxstr_savelen (sxstrmngr_t *strtable, char *string, SXUINT strlength)
{
    sxstrtable_t	*ptable, *qtable;
    SXINT	ste;
    SXUINT	hash;

    if ((ste = strtable->head [hash = sxscrmbl (string, strlength) % SXHASH_LENGTH]) == SXERROR_STE)
	/* Brand new string */
	return strtable->head [hash] = sxstr_add (strtable, string, strlength);

    qtable = NULL;

    do {
	ptable = &(strtable->table [ste]);

	if (ptable->length == strlength &&
	    (strlength == 0 || (*string == *(ptable->pointer) &&
				sxstrequal (string, ptable->pointer, strlength)))) {
	    /* string already exists */
	    if (qtable != NULL) {
		/* Arrange to find it faster next time */
		qtable->collision_link = ptable->collision_link;
		ptable->collision_link = strtable->head [hash];
		strtable->head [hash] = ste;
	    }

	    return ste;
	}
    } while ((ste = (qtable = ptable)->collision_link) != SXERROR_STE);

    /* We must do this in two steps, as sxstr_add may reallocate table */

    ptable->collision_link = strtable->top;
    return sxstr_add (strtable, string, strlength);
}

/*---------------------------------------------------------------------------*/

#ifdef __GNUC__

/* sxstr_retrieve() is defined using the GNUC-specific extension ({ ... }) */

#else

SXINT sxstr_save (sxstrmngr_t *strtable, char *string)
{
     char *strvar;

     strvar = string;
     return sxstr_savelen (strtable, strvar, strlen (strvar));
}

#endif

/*---------------------------------------------------------------------------*/

/* Searches for the existence of string in strtable; return the unique identifier of string or 0 if it does not exist */

SXINT sxstr_retrievelen (sxstrmngr_t *strtable, char *string, SXUINT strlength)
{
    sxstrtable_t	*ptable;
    SXINT	ste;

    if ((ste = strtable->head [sxscrmbl (string, strlength) % SXHASH_LENGTH]) != SXERROR_STE)
	do {
	    ptable = &(strtable->table [ste]);

	    if (ptable->length == strlength && sxstrequal (string, ptable->pointer, strlength)) {
		/* string already exists */
		break;
	    }
	} while ((ste = ptable->collision_link) != SXERROR_STE);

    return ste;
}

/*---------------------------------------------------------------------------*/

#ifdef __GNUC__

/* sxstr_retrieve() is defined using the GNUC-specific extension ({ ... }) */

#else

SXINT sxstr_retrieve (sxstrmngr_t *strtable, char *string)
{
     char *strvar;

     strvar = string;
     return sxstr_retrievelen (strtable, strvar, strlen (strvar));
}

#endif

/*---------------------------------------------------------------------------*/

static void free_strings (sxstrstring_t *string)
{
    while (string != NULL) {
	sxstrstring_t	*next = string->next;

	sxfree (string);
	string = next;
    }
}

/*---------------------------------------------------------------------------*/

static void sxstr_create (sxstrmngr_t *strtable)
{
    if (strtable->head == NULL) {
	/* Allocate */
	strtable->head = (SXINT*) sxcalloc (SXHASH_LENGTH, sizeof (SXINT));
	strtable->table = (sxstrtable_t*) sxalloc (strtable->tablength = BUFSIZ, sizeof (sxstrtable_t));
	strtable->string = (sxstrstring_t*) sxalloc (1, sizeof (sxstrstring_t*) + (strtable->strlength = BUFSIZ*4 - sizeof (sxstrstring_t*)) * sizeof (char));
    }
    else
	free_strings (strtable->string->next);
    
    strtable->string->next = NULL;
    
    if (SXERROR_STE != 0 || strtable->top != 0) {
	/* Initialise */
	SXINT	*phead;
	
	for (phead = &(strtable->head [SXHASH_LENGTH - 1]); phead >= strtable->head; *phead-- = SXERROR_STE) {
	    /* null; */
	}
    }

    /* On range des chaines vides pour "SXERROR_STE" et "SXEMPTY_STE".  */

    *(strtable->table [SXERROR_STE].pointer = strtable->string->chars) = SXNUL;
    strtable->table [SXERROR_STE].length = 0;
    strtable->table [SXERROR_STE].collision_link = SXERROR_STE;
    strtable->table [SXEMPTY_STE] = strtable->table [SXERROR_STE];
    strtable->head [0] = SXEMPTY_STE;
    strtable->top = SXEMPTY_STE + 1;
    strtable->strnext = 1;
}

/*---------------------------------------------------------------------------*/

static void sxstr_delete (sxstrmngr_t *strtable)
{
    free_strings (strtable->string), strtable->string = NULL;
    sxfree (strtable->table), strtable->table = NULL;
    sxfree (strtable->head), strtable->head = NULL;
}

/*---------------------------------------------------------------------------*/

void sxstr_mngr (SXINT sxstr_mngr_what, ...)
/* si sxstr_mngr_what vaut SXOPEN ou SXCLOSE, on a un second parametre 
 * strtable de type (sxstrmngr_t *) */
{
    va_list ap;
    sxstrmngr_t *strtable;

    switch (sxstr_mngr_what) {
    case SXBEGIN:
	/* creation de la table par defaut */
    case SXCLEAR:
	/* remise a zero de la table par defaut */
	sxstr_create (&sxstrmngr);
	break;

    case SXEND:
	/* destruction de la table par defaut */
	sxstr_delete (&sxstrmngr);
	break;

    case SXOPEN:
	/* creation d'une table supplementaire voulue par l'utilisateur */
	va_start (ap, sxstr_mngr_what);
	strtable = va_arg (ap, sxstrmngr_t*);
	sxstr_create (strtable);
	va_end (ap);
	break;

    case SXCLOSE:
	/* destruction d'une table supplementaire voulue par l'utilisateur */
	va_start (ap, sxstr_mngr_what);
	strtable = va_arg (ap, sxstrmngr_t*);
	sxstr_delete (strtable);
	va_end (ap);
	break;

    default: /* pour faire taire gcc -Wswitch-default */
#if EBUG
      sxtrap("sxstr_mngr","unknown switch case #1");
#endif
      break;
    }
}

/*---------------------------------------------------------------------------*/

void sxstr_dump (FILE *f, sxstrmngr_t *strtable)
{
    SXINT n, i;

    n = strtable->top;
    fprintf (f, "contents of string table at 0x%08lx (%ld elements)\n", 
        (SXINT) strtable, n);
    for (i = 0 ; i < n; i++) {
        fprintf (f, "\tentry %ld -> \"%s\" (length = %lu)\n", 
	    i, sxstr_get (strtable, i), sxstr_length (strtable, i));
    }
}

/*---------------------------------------------------------------------------*/

void sxstr_save_keywords (sxstrmngr_t *strtable, char *lang)
{
    if (strcmp (lang, "C") == 0) {
	sxstr_save (strtable, "asm");
	sxstr_save (strtable, "auto");
	sxstr_save (strtable, "bool"); /* not a keyword, strictly speaking */
	sxstr_save (strtable, "break");
	sxstr_save (strtable, "case");
	sxstr_save (strtable, "char");
	sxstr_save (strtable, "const");
	sxstr_save (strtable, "continue");
	sxstr_save (strtable, "default");
	sxstr_save (strtable, "do");
	sxstr_save (strtable, "double");
	sxstr_save (strtable, "else");
	sxstr_save (strtable, "enum");
	sxstr_save (strtable, "extern");
	sxstr_save (strtable, "false"); /* not a keyword, strictly speaking */
	sxstr_save (strtable, "float");
	sxstr_save (strtable, "for");
	sxstr_save (strtable, "fortran");
	sxstr_save (strtable, "goto");
	sxstr_save (strtable, "if");
	sxstr_save (strtable, "inline");
	sxstr_save (strtable, "int");
	sxstr_save (strtable, "long");
	sxstr_save (strtable, "main"); /* not a keyword, strictly speaking */
	sxstr_save (strtable, "register");
	sxstr_save (strtable, "restrict");
	sxstr_save (strtable, "return");
	sxstr_save (strtable, "short");
	sxstr_save (strtable, "signed");
	sxstr_save (strtable, "sizeof");
	sxstr_save (strtable, "static");
	sxstr_save (strtable, "struct");
	sxstr_save (strtable, "switch");
	sxstr_save (strtable, "true"); /* not a keyword, strictly speaking */
	sxstr_save (strtable, "typedef");
	sxstr_save (strtable, "union");
	sxstr_save (strtable, "unsigned");
	sxstr_save (strtable, "void");
	sxstr_save (strtable, "volatile");
	sxstr_save (strtable, "while");
	sxstr_save (strtable, "_Alignas");
	sxstr_save (strtable, "_Alignof");
	sxstr_save (strtable, "_Atomic");
	sxstr_save (strtable, "_Bool");
	sxstr_save (strtable, "_Complex");
	sxstr_save (strtable, "_Generic");
	sxstr_save (strtable, "_Imaginary");
	sxstr_save (strtable, "_Noreturn");
	sxstr_save (strtable, "_Pragma");
	sxstr_save (strtable, "_Static_assert");
	sxstr_save (strtable, "_Thread_local");
    } else if (strcmp (lang, "LOTOS") == 0) {
        /* reserved "normal" identifiers */
	sxstr_save (strtable, "accept");
	sxstr_save (strtable, "actualizedby");
	sxstr_save (strtable, "any");
	sxstr_save (strtable, "behavior");
	sxstr_save (strtable, "behaviour");
	sxstr_save (strtable, "choice");
	sxstr_save (strtable, "endlib");
	sxstr_save (strtable, "endproc");
	sxstr_save (strtable, "endspec");
	sxstr_save (strtable, "endtype");
	sxstr_save (strtable, "eqns");
	sxstr_save (strtable, "exit");
	sxstr_save (strtable, "for");
	sxstr_save (strtable, "forall");
	sxstr_save (strtable, "formaleqns");
	sxstr_save (strtable, "formalopns");
	sxstr_save (strtable, "formalsorts");
	sxstr_save (strtable, "hide");
	sxstr_save (strtable, "i");
	sxstr_save (strtable, "in");
	sxstr_save (strtable, "is");
	sxstr_save (strtable, "let");
	sxstr_save (strtable, "library");
	sxstr_save (strtable, "noexit");
	sxstr_save (strtable, "of");
	sxstr_save (strtable, "ofsort");
	sxstr_save (strtable, "opnnames");
	sxstr_save (strtable, "opns");
	sxstr_save (strtable, "par");
	sxstr_save (strtable, "process");
	sxstr_save (strtable, "renamedby");
	sxstr_save (strtable, "sortnames");
	sxstr_save (strtable, "sorts");
	sxstr_save (strtable, "specification");
	sxstr_save (strtable, "stop");
	sxstr_save (strtable, "type");
	sxstr_save (strtable, "using");
	sxstr_save (strtable, "where");
	sxstr_save (strtable, "_");
        /* reserved "special" identifiers */
	sxstr_save (strtable, "=");
	sxstr_save (strtable, "=>");
	sxstr_save (strtable, "->");
	sxstr_save (strtable, ">>");
    } else {
	 /* unsupported language : do nothing */
    }
}

/*---------------------------------------------------------------------------*/

